### PR TITLE
fixed #16 - Implement progress bar for subtree processing

### LINKED ISSUE #16: Loading subtree issues may take some time, we need a loading indicator
The accessibility panel in Firefox DevTools has one when you click on one of the "checks" button (color contrast for example). We could draw inspiration from this.

### PR DESCRIPTION
**Pull request by dadaa**

### CODE REVIEW

File: extension/sidebar-pane.js
Comment (captainbrosset) [inline extension/sidebar-pane.js:53]: Maybe rephrase to `Getting all descendants of the selected node`.

File: extension/experiments/inspectedNode/api-client.js
Comment (captainbrosset) [inline extension/experiments/inspectedNode/api-client.js:44]: Can you please add a bit of code comments here to explain what is going on? 
This progress listener code is quite complicated, and a comment would help a lot.

File: extension/experiments/inspectedNode/api-client.js
Comment (captainbrosset) [inline extension/experiments/inspectedNode/api-client.js:64]: Same here.

File: extension/progress-bar.js
Comment (captainbrosset) [inline extension/progress-bar.js:22]: Just `increment` would be fine.

File: extension/progress-bar.js
Comment (captainbrosset) [inline extension/progress-bar.js:3]: Please add a short jsdoc comment that explains what `parentEl`, `labelEl` and `progressEl` should be.
And also how to use this class. What do `enable` and `disable` do, etc.

File: extension/sidebar-pane.js
Comment (captainbrosset) [inline extension/sidebar-pane.js:58]: Why `+2` here?

File: extension/sidebar-pane.js
Comment (captainbrosset) [inline extension/sidebar-pane.js:61]: Maybe rephrase to `Getting web compatibility issues for elements`.

File: extension/sidebar-pane.js
Comment (captainbrosset) [inline extension/sidebar-pane.js:78]: `Getting the applied CSS styles for all elements`

File: extension/sidebar-pane.js
Comment (captainbrosset) [inline extension/sidebar-pane.js:82]: `Getting web compatibility issues for CSS styles`.

File: extension/sidebar-pane.js
Comment (captainbrosset) [inline extension/sidebar-pane.js:90]: `Rendering all issues`.

**Review by captainbrosset — CHANGES_REQUESTED** [review]
Wow, that is pretty impressive to read. I was not expecting the progress bar to actually move accurately! 😱 

There are a few reasons why I think we might *not* want to do this though:

* this code is quite complex to understand at first sight (maybe not hard to maintain though, so that might be ok)
* maybe my laptop is just too powerful, but it's hard to see the progress bar for more than an instant
* the code is inside the experimental API which is something that only works on Firefox and not in other browsers.

So, because of these reasons I would be tempted to add an infinite spinner instead of a progress bar.
With a spinner everything becomes much simpler. No experimental API code to write. You just start it when a node is selected, and you remove it when the warnings are returned.

Now, I do like the accuracy of this progress bar, and on lower-end machines, it might actually become more important for it to be displayed like this.

The other counter-argument to what I said is: on other browsers that don't support this API, we could always show a different loader, and only have this one on Firefox.

So I'm not against landing this per say. I just think that we could achieve a similar result with way less code 😄

**Comment by oslego** [comment]
First, let me say that this is awesome! I really appreciate your diligence with showing an accurate progress indicator, @dadaa!

But after carefully reading the code, I echo Patrick's concerns.

The progress indicator is fantastic, but on our machines (which are quite powerful, yes) the indicator is barely visible. Any meaningful delay would come from inspecting a very complex and very deeply nested DOM due to the recursive child traversal.

Fortunately, [we have a Talos test](https://searchfox.org/mozilla-central/source/testing/talos/talos/tests/devtools/addon/content/pages/custom/inspector/index.html) for this sort of scenario to check DevTools performance (you can download the file and run it on a local server). It generates 50 levels of nesting under the first node. That's pretty extreme, but even in this scenario, the action is almost instant and the indicator is barely visible. 

As Patrick says, it may be more beneficial on slower machines, but I'd argue in that case that any additional computational overhead is adding yet more slowness to the process with the event handling and DOM manipulation for the progress indicator itself.

Given the added complexity and the hard dependency on the experimental API (which we may not be able to use if we port this extension to another browser), I agree with Patrick that a sensible approach for ourselves and future maintainers would be to take the easy route and use an indeterminate loading spinner (achievable perhaps still with a `<progress>` element and the [`:indeterminate` CSS pseudo-class](https://developer.mozilla.org/en-US/docs/Web/CSS/:indeterminate))

I can understand you worked hard on this and you're proud of it. You should be, it's amazing!
 
But I encourage keeping code complexity low so we have room in future developers' heads so they can grasp what's going on with the business logic without getting confused by the progress indicator which, in this case, while great, it seems to be mostly invisible to users on reasonably-powered machines.

My suggestion would be to not merge this patch, but still keep it around. If we discover that there are viable cases where a highly accurate progress indicator is required for categorically improving user experience, then we should use it and accept the complexity then.

**Comment by dadaa** [comment]
Thank you so much @captainbrosset and @oslego !
Yep, I understand your concern. 
I'll keep this PR as it is and create one more PR which fixes by another simple way.

**Pull request closed without merging**
